clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsGroupHeartbeatRequestManager.java

```diff
@@ -0,0 +1,387 @@
+            data.setMemberId(membershipManager.memberId());
+            data.setMemberEpoch(membershipManager.memberEpoch());
+            membershipManager.groupInstanceId().ifPresent(data::setInstanceId);
+            StreamsGroupHeartbeatRequestData.Topology topology = new StreamsGroupHeartbeatRequestData.Topology();
+            topology.setSubtopologies(getTopologyFromStreams());
+            topology.setEpoch(streamsRebalanceData.topologyEpoch());
+            data.setRebalanceTimeoutMs(rebalanceTimeoutMs);
+            data.setTopology(topology);
+            data.setProcessId(streamsRebalanceData.processId().toString());
+            data.setActiveTasks(Collections.emptyList());
+            data.setStandbyTasks(Collections.emptyList());
+            data.setWarmupTasks(Collections.emptyList());
+            streamsRebalanceData.endpoint().ifPresent(userEndpoint -> {
+                data.setUserEndpoint(new StreamsGroupHeartbeatRequestData.Endpoint()
+                    .setHost(userEndpoint.host())
+                    .setPort(userEndpoint.port())
+                );
+            });
+            data.setClientTags(streamsRebalanceData.clientTags().entrySet().stream()
+                .map(entry -> new StreamsGroupHeartbeatRequestData.KeyValue()
+                    .setKey(entry.getKey())
+                    .setValue(entry.getValue())
+                )
+                .collect(Collectors.toList()));
+            if (streamsRebalanceData.shutdownRequested()) {
+                data.setShutdownApplication(true);
+            }
+            StreamsRebalanceData.Assignment reconciledAssignment = streamsRebalanceData.reconciledAssignment();
+            data.setActiveTasks(convertTaskIdCollection(reconciledAssignment.activeTasks()));
+            data.setStandbyTasks(convertTaskIdCollection(reconciledAssignment.standbyTasks()));
+            data.setWarmupTasks(convertTaskIdCollection(reconciledAssignment.warmupTasks()));
+            return data;
+        }
+
+        private static List<StreamsGroupHeartbeatRequestData.TaskIds> convertTaskIdCollection(final Set<StreamsRebalanceData.TaskId> tasks) {
+            return tasks.stream()
+                .collect(
+                    Collectors.groupingBy(StreamsRebalanceData.TaskId::subtopologyId,
+                        Collectors.mapping(StreamsRebalanceData.TaskId::partitionId, Collectors.toList()))
+                )
+                .entrySet()
+                .stream()
+                .map(entry -> {
+                    StreamsGroupHeartbeatRequestData.TaskIds ids = new StreamsGroupHeartbeatRequestData.TaskIds();
+                    ids.setSubtopologyId(entry.getKey());
+                    ids.setPartitions(entry.getValue());
+                    return ids;
+                })
+                .collect(Collectors.toList());
+        }
+
+        private List<StreamsGroupHeartbeatRequestData.Subtopology> getTopologyFromStreams() {
```

Comment:
Would seem cleaner to me to make this a static method and pass `streamsRebalanceData.subtopologies()` as a parameter.

Reply:
Done